test/e2e/clusters/clusters.go

```diff
@@ -0,0 +1,256 @@
+		return distroToUse, nil
+	} else {
+		return DistroUnknown, fmt.Errorf("\"%v\" is not a valid value for determining which k8s distro to use", s)
+	}
+}
+
+// CreateClusterWithTemporaryKubeconfig creates a temporary Kubeconfig file, updates the process's KUBECONFIG env var
+// to point at the new file, and creates a kubernetes cluster. It returns the path to the temporary Kubeconfig file,
+// or an error if something went wrong. The caller is responsible for destroying the cluster and deleting the temporary
+// kubeconfig file. The KUBECONFIG env var doesn't need to be reset since it was only changed for the current process.
+func CreateClusterWithTemporaryKubeconfig(distroToUse DistroToUse) (string, error) {
+	// Create the temporary Kubeconfig file
+	var clusterName string
+	switch distroToUse {
+	case DistroKind:
+		clusterName = kindClusterName
+	case DistroK3d:
+		clusterName = k3dClusterName
+	}
+	tempKubeconfigFile, err := ioutil.TempFile("", clusterName+"*.yaml")
+	if err != nil {
+		return "", err
+	}
+
+	// Set the KUBECONFIG env var to the new file
+	err = os.Setenv("KUBECONFIG", tempKubeconfigFile.Name())
+	if err != nil {
+		return "", err
+	}
+
+	// Create the cluster
+	switch distroToUse {
+	case DistroKind:
+		err = createKindCluster(tempKubeconfigFile.Name())
+		if err != nil {
+			return "", err
+		}
+	case DistroK3d:
+		err = createK3dClusterUsingCurrentKubeconfig()
+		if err != nil {
+			return "", err
+		}
+	}
+	return tempKubeconfigFile.Name(), nil
+}
+
+// DeleteClusterAndTemporaryKubeconfig is intended to be deferred at the end of TestMain execution so we can clean up
+// the cluster and temp file that we made.
+func DeleteClusterAndTemporaryKubeconfig(distroToUse DistroToUse, tempKubeconfigFilePath string) error {
+	// Defer the temp file delete so it always happens no matter what
+	defer func(name string) {
+		_ = os.Remove(name)
+	}(tempKubeconfigFilePath)
+
+	// Delete the cluster
+	switch distroToUse {
+	case DistroKind:
+		err := deleteKindCluster(tempKubeconfigFilePath)
+		if err != nil {
+			return err
+		}
+	case DistroK3d:
+		err := deleteK3dCluster()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// TryValidateClusterIsRunning establishes a valid connection to the currently configured cluster, then returns,
+// throwing an error if anything went wrong.
+func TryValidateClusterIsRunning() error {
+	clientSet, err := GetClientSet()
+	if err != nil {
+		return fmt.Errorf("unable to connect to the cluster: %w", err)
+	}
+	pods, err := clientSet.CoreV1().Pods("kube-system").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to list running pods: %w", err)
+	}
+	if len(pods.Items) < 1 {
+		return fmt.Errorf("established connection to cluster, but no pods were found")
+	}
+
+	return nil
+}
+
+// GetConfig gets the currently configured Kubeconfig, or throws an error if something went wrong.
+func GetConfig() (*rest.Config, error) {
```

Comment:
Concerned about these functions being similar + similar name to the k8s package, is there a reason they had to be custom logic here vs use the core code?  We don't want custom connect logic for tests because then we're cheating the E2E tests.

Reply:
This doesn't cheat the E2E tests, they still use the golang application code to do its thing, through the Zarf binary that gets used.

This gets used in the TestMain function to set up the cluster.

Reply:
Copy so I guess a better question is how does this differ from the same functions in the k8s package?